Intermittent test failure: TestNimbusExperimentSerializer.test_serializers_with_feature_value_None

Seen in https://app.circleci.com/pipelines/github/mozilla/experimenter/8986/workflows/bed04c8b-4541-460f-95b2-cb2368cecefe/jobs/24769

```

=================================== FAILURES ===================================
___ TestNimbusExperimentSerializer.test_serializers_with_feature_value_None ____
[gw1] linux -- Python 3.9.2 /usr/local/bin/python3

schema_name = 'experiments/NimbusExperiment'
instance = {'schemaVersion': '1.4.0', 'slug': 'decentralized-intangible-capability', 'id': 'decentralized-intangible-capability',...llment': 2, 'referenceBranch': 'down-sized-actuating-adapter', 'featureIds': ['re-engineered-24hour-customer-loyalty']}

    def check_schema(schema_name: str, instance: any) -> bool:
        """
        Check if `instance` matches the schema named `schema_name`.
    
        Returns `True` if it does. If it does not match the schema,
        `mozilla_nimbus_shared.NimbusValidationError` is thrown. It contains an
        attribute `.errors` that detail all the ways the data does not match the
        schema.
    
        If the schema name given does not refer to a known schema, then a
        `mozilla_nimbus_shared.NimbusSchemaNotFoundError` exception is thrown.
        """
    
        if schema_name not in _validator_cache:
            schema_path = "schemas/" + schema_name + ".json"
            schema_text = pkgutil.get_data("mozilla_nimbus_shared", schema_path)
            schema = json.loads(schema_text)
            _validator_cache[schema_name] = jsonschema.Draft7Validator(schema)
    
        validator = _validator_cache[schema_name]
        try:
>           validator.validate(instance)

/usr/local/lib/python3.9/site-packages/mozilla_nimbus_shared/__init__.py:43: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <jsonschema.validators.create.<locals>.Validator object at 0x7f4715d729a0>
args = ({'schemaVersion': '1.4.0', 'slug': 'decentralized-intangible-capability', 'id': 'decentralized-intangible-capability'...ment': 2, 'referenceBranch': 'down-sized-actuating-adapter', 'featureIds': ['re-engineered-24hour-customer-loyalty']},)
kwargs = {}
error = <ValidationError: "'randomizationUnit' is a required property">

    def validate(self, *args, **kwargs):
        for error in self.iter_errors(*args, **kwargs):
>           raise error
E           jsonschema.exceptions.ValidationError: 'randomizationUnit' is a required property
E           
E           Failed validating 'required' in schema['properties']['bucketConfig']:
E               {'description': 'Bucketing configuration',
E                'properties': {'count': {'description': 'Number of buckets to check',
E                                         'type': 'integer'},
E                               'namespace': {'description': 'Additional inputs to the '
E                                                            'hashing function',
E                                             'type': 'string'},
E                               'randomizationUnit': {'description': 'A unique, stable '
E                                                                    'identifier for '
E                                                                    'the user used as '
E                                                                    'an input to '
E                                                                    'bucket hashing',
E                                                     'type': 'string'},
E                               'start': {'description': 'Index of start of the range '
E                                                        'of buckets',
E                                         'type': 'integer'},
E                               'total': {'default': 10000,
E                                         'description': 'Total number of buckets. You '
E                                                        'can assume this will always '
E                                                        'be 10000.',
E                                         'type': 'integer'}},
E                'required': ['randomizationUnit',
E                             'namespace',
E                             'start',
E                             'count',
E                             'total'],
E                'type': 'object'}
E           
E           On instance['bucketConfig']:
E               OrderedDict([('namespace', 'decentralized-intangible-capability-1'),
E                            ('start', 0),
E                            ('count', 8000),
E                            ('total', 10000)])

/usr/local/lib/python3.9/site-packages/jsonschema/validators.py:353: ValidationError

During handling of the above exception, another exception occurred:

self = <experimenter.experiments.tests.api.v6.test_serializers.TestNimbusExperimentSerializer testMethod=test_serializers_with_feature_value_None>

    def test_serializers_with_feature_value_None(self):
        experiment = NimbusExperimentFactory.create_with_status(
            NimbusExperiment.Status.ACCEPTED,
            branches=[],
        )
        experiment.reference_branch = NimbusBranchFactory(
            experiment=experiment, feature_value=None
        )
        experiment.save()
        serializer = NimbusExperimentSerializer(experiment)
        self.assertIsNone(serializer.data["branches"][0]["feature"]["value"])
>       check_schema("experiments/NimbusExperiment", serializer.data)

experimenter/experiments/tests/api/v6/test_serializers.py:110: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

schema_name = 'experiments/NimbusExperiment'
instance = {'schemaVersion': '1.4.0', 'slug': 'decentralized-intangible-capability', 'id': 'decentralized-intangible-capability',...llment': 2, 'referenceBranch': 'down-sized-actuating-adapter', 'featureIds': ['re-engineered-24hour-customer-loyalty']}

    def check_schema(schema_name: str, instance: any) -> bool:
        """
        Check if `instance` matches the schema named `schema_name`.
    
        Returns `True` if it does. If it does not match the schema,
        `mozilla_nimbus_shared.NimbusValidationError` is thrown. It contains an
        attribute `.errors` that detail all the ways the data does not match the
        schema.
    
        If the schema name given does not refer to a known schema, then a
        `mozilla_nimbus_shared.NimbusSchemaNotFoundError` exception is thrown.
        """
    
        if schema_name not in _validator_cache:
            schema_path = "schemas/" + schema_name + ".json"
            schema_text = pkgutil.get_data("mozilla_nimbus_shared", schema_path)
            schema = json.loads(schema_text)
            _validator_cache[schema_name] = jsonschema.Draft7Validator(schema)
    
        validator = _validator_cache[schema_name]
        try:
            validator.validate(instance)
        except jsonschema.exceptions.ValidationError:
>           raise NimbusValidationError(validator.iter_errors(instance))
E           mozilla_nimbus_shared.NimbusValidationError: Data does not match schema

/usr/local/lib/python3.9/site-packages/mozilla_nimbus_shared/__init__.py:45: NimbusValidationError
```

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/EXP-1157)
┆Issue Number: EXP-1157

--- a/app/experimenter/experiments/constants/nimbus.py
+++ b/app/experimenter/experiments/constants/nimbus.py
@@ -101,6 +101,11 @@ class Channel(models.TextChoices):
     RELEASE = "release"
 
 
+class BucketRandomizationUnit(models.TextChoices):
+    NORMANDY = "normandy_id"
+    NIMBUS = "nimbus_id"
+
+
 @dataclass
 class ApplicationConfig:
     name: str
@@ -109,6 +114,7 @@ class ApplicationConfig:
     channel_app_id: Dict[str, str]
     default_app_id: str
     collection: str
+    randomization_unit: str
 
 
 APPLICATION_CONFIG_DESKTOP = ApplicationConfig(
@@ -122,6 +128,7 @@ APPLICATION_CONFIG_DESKTOP = ApplicationConfig(
     },
     default_app_id="firefox-desktop",
     collection=settings.KINTO_COLLECTION_NIMBUS_DESKTOP,
+    randomization_unit=BucketRandomizationUnit.NORMANDY,
 )
 
 APPLICATION_CONFIG_FENIX = ApplicationConfig(
@@ -135,6 +142,7 @@ APPLICATION_CONFIG_FENIX = ApplicationConfig(
     },
     default_app_id="",
     collection=settings.KINTO_COLLECTION_NIMBUS_MOBILE,
+    randomization_unit=BucketRandomizationUnit.NIMBUS,
 )
 
 APPLICATION_CONFIG_IOS = ApplicationConfig(
@@ -148,6 +156,7 @@ APPLICATION_CONFIG_IOS = ApplicationConfig(
     },
     default_app_id="",
     collection=settings.KINTO_COLLECTION_NIMBUS_MOBILE,
+    randomization_unit=BucketRandomizationUnit.NIMBUS,
 )
 
 
@@ -300,15 +309,6 @@ class NimbusConstants(object):
 
     class EmailType(models.TextChoices):
         EXPERIMENT_END = "experiment end"
-
-    class BucketRandomizationUnit(models.TextChoices):
-        NORMANDY = "normandy_id"
-        NIMBUS = "nimbus_id"
-
-    APPLICATION_BUCKET_RANDOMIZATION_UNIT = {
-        Application.DESKTOP: BucketRandomizationUnit.NORMANDY,
-        Application.FENIX: BucketRandomizationUnit.NIMBUS,
-    }
 
     EMAIL_EXPERIMENT_END_SUBJECT = "Action required: Please turn off your Experiment"
 

--- a/app/experimenter/experiments/models/nimbus.py
+++ b/app/experimenter/experiments/models/nimbus.py
@@ -332,7 +332,7 @@ class NimbusIsolationGroup(models.Model):
 
     @property
     def randomization_unit(self):
-        return NimbusExperiment.APPLICATION_BUCKET_RANDOMIZATION_UNIT[self.application]
+        return NimbusExperiment.APPLICATION_CONFIGS[self.application].randomization_unit
 
     @property
     def namespace(self):

--- a/app/experimenter/experiments/tests/api/v6/test_serializers.py
+++ b/app/experimenter/experiments/tests/api/v6/test_serializers.py
@@ -96,9 +96,11 @@ class TestNimbusExperimentSerializer(TestCase):
 
         check_schema("experiments/NimbusExperiment", serializer.data)
 
-    def test_serializers_with_feature_value_None(self):
+    @parameterized.expand(list(NimbusExperiment.Application))
+    def test_serializers_with_feature_value_None(self, application):
         experiment = NimbusExperimentFactory.create_with_status(
             NimbusExperiment.Status.ACCEPTED,
+            application=application,
             branches=[],
         )
         experiment.reference_branch = NimbusBranchFactory(

--- a/app/experimenter/experiments/tests/test_models/test_models_nimbus.py
+++ b/app/experimenter/experiments/tests/test_models/test_models_nimbus.py
@@ -494,19 +494,7 @@ class TestNimbusDocumentLink(TestCase):
         self.assertEqual(str(doco_link), f"{doco_link.title} ({doco_link.link})")
 
 
-@parameterized_class(
-    ("application", "randomization_unit"),
-    [
-        [
-            NimbusExperiment.Application.DESKTOP,
-            NimbusExperiment.BucketRandomizationUnit.NORMANDY,
-        ],
-        [
-            NimbusExperiment.Application.FENIX,
-            NimbusExperiment.BucketRandomizationUnit.NIMBUS,
-        ],
-    ],
-)
+@parameterized_class(("application",), [list(NimbusExperiment.Application)])
 class TestNimbusIsolationGroup(TestCase):
     def test_empty_isolation_group_creates_isolation_group_and_bucket_range(self):
         """
@@ -526,7 +514,7 @@ class TestNimbusIsolationGroup(TestCase):
         self.assertEqual(bucket.isolation_group.total, NimbusExperiment.BUCKET_TOTAL)
         self.assertEqual(
             bucket.isolation_group.randomization_unit,
-            self.randomization_unit,
+            experiment.application_config.randomization_unit,
         )
 
     def test_existing_isolation_group_adds_bucket_range(self):
@@ -548,7 +536,7 @@ class TestNimbusIsolationGroup(TestCase):
         self.assertEqual(bucket.isolation_group, isolation_group)
         self.assertEqual(
             bucket.isolation_group.randomization_unit,
-            self.randomization_unit,
+            experiment.application_config.randomization_unit,
         )
 
     def test_existing_isolation_group_with_buckets_adds_next_bucket_range(self):
@@ -573,7 +561,7 @@ class TestNimbusIsolationGroup(TestCase):
         self.assertEqual(bucket.isolation_group, isolation_group)
         self.assertEqual(
             bucket.isolation_group.randomization_unit,
-            self.randomization_unit,
+            experiment.application_config.randomization_unit,
         )
 
     def test_full_isolation_group_creates_next_isolation_group_adds_bucket_range(
@@ -603,7 +591,7 @@ class TestNimbusIsolationGroup(TestCase):
         self.assertEqual(bucket.isolation_group.instance, isolation_group.instance + 1)
         self.assertEqual(
             bucket.isolation_group.randomization_unit,
-            self.randomization_unit,
+            experiment.application_config.randomization_unit,
         )
 
     def test_existing_isolation_group_with_matching_name_but_not_application_is_filtered(

